### PR TITLE
osd/OSD: introduce reset_purged_snaps_last

### PR DESCRIPTION
**Pull request by Matan-B**
```
When the OSD preboots it sends a MMonGetPurgedSnaps message to
the monitor (`_get_purged_snaps`).
The monitor will reply with all the purged snapshots that their purged_epoch_ is in the
range of superblock.purged_snaps_last + 1 up to the last superblock.current_epoch + 1.
When the OSD will handle the reply from the mon (`handle_get_purged_snaps_reply`)
it will call `record_purged_snaps` to write those purged snapshots in the
OSD store as well (PSN_ keys).

Once purged_snaps_last is reset, in the following OSD reboot, *all* the snapshots that were marked
as purged (purged_snaps_ keys) in the mon's store will be also marked,
correspondingly, in the OSD store.
That way `scrub_purged_snaps` will be able to re-trim the snapshots that weren't
marked as purged in the OSD side (for some reason)

Fixes: https://tracker.ceph.com/issues/62981
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by athanatos — APPROVED** [review]
Yep, this seems like a reasonable fix.

**Comment by athanatos** [comment]
Oh, if the monitors have already trimmed maps (and, in general, they will have), do the mons send all purged snaps anyway?

**Comment by Matan-B** [comment]
> Oh, if the monitors have already trimmed maps (and, in general, they will have), do the mons send all purged snaps anyway?

Yes, mon's `purged_snap_` keys are not trimmed.

**Comment by idryomov** [comment]
> Yes, mon's `purged_snap_` keys are not trimmed.

There was a thread on the dev list recently where this came up [1].  Ignore the rbd-mirror discussion there -- it was only tangentially related.  I'm linking it to highlight that we seem to have recurring issues in the snapshot trimming area: OSDs aren't picking up purged snap IDs from the monitors [2], monitor db growing many millions of osd_snap/purged_snap_ key which causes monitor startup slowness, etc.

Manually resetting `superblock.purged_snaps_last` should help with clearing up the backlog, but do we know where it's coming from to begin with?

[1] https://lists.ceph.io/hyperkitty/list/dev@ceph.io/message/UOJG46YXTIPOXJUSELIN42ATAD5FPMDY/
[2] https://bugzilla.redhat.com/show_bug.cgi?id=2223533

**Comment by rzarzynski** [comment]
We should be able to backport this, so ticket + `Fixes` are needed.

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -3126,6 +3126,16 @@ will start to track new ops received afterwards.";
...
   }
 
+  else if (prefix == "reset_purged_snaps_last") {
+    lock_guard l(osd_lock);
+    superblock.purged_snaps_last = 0;
+    ObjectStore::Transaction t;
+    dout(10) << __func__ << " updating superblock" << dendl;
+    write_superblock(cct, superblock, t);
+    int tr = store->queue_transaction(service.meta_ch, std::move(t), nullptr);
+    ceph_assert(tr == 0);
```
[the remark this answers is not in the document]
Comment (Matan-B) [reply src/osd/OSD.cc]: Looks like not crashing here is preferable. Updated, Thanks.

**Comment by Matan-B** [comment]
> There was a thread on the dev list recently where this came up [1]. Ignore the rbd-mirror discussion there -- it was only tangentially related. I'm linking it to highlight that we seem to have recurring issues in the snapshot trimming area: OSDs aren't picking up purged snap IDs from the monitors [2], monitor db growing many millions of osd_snap/purged_snap_ key which causes monitor startup slowness, etc.
> 
> Manually resetting `superblock.purged_snaps_last` should help with clearing up the backlog, but do we know where it's coming from to begin with?
> 
> [1] https://lists.ceph.io/hyperkitty/list/dev@ceph.io/message/UOJG46YXTIPOXJUSELIN42ATAD5FPMDY/ [2] https://bugzilla.redhat.com/show_bug.cgi?id=2223533

Thanks for sharing the thread.
The 2 issues shared aren't related. As to the one reported in the thread, I agree we should have a solution in hand. Simply trimming the old keys is probably possible. However, the fact that we never trimmed those keys proved that it may be useful for cases such as the one in the BZ attached.
We should probably be able to better compress/merge contiguous keys *also* after they are stored.  I'll open a separate tracker for it as this is out of the scope of this PR.

**Comment by idryomov** [comment]
> We should probably be able to better compress/merge contiguous keys _also_ after they are stored. I'll open a separate tracker for it as this is out of the scope of this PR.

Sounds good.

What is in scope of this PR is the issue of OSDs not picking up purged snap IDs from the monitors.  Do we know what is causing that?

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -3126,6 +3126,19 @@ will start to track new ops received afterwards.";
...
 
+  else if (prefix == "reset_purged_snaps_last") {
+    lock_guard l(osd_lock);
+    superblock.purged_snaps_last = 0;
+    ObjectStore::Transaction t;
+    dout(10) << __func__ << " updating superblock" << dendl;
+    write_superblock(cct, superblock, t);
+    int ret = store->queue_transaction(service.meta_ch, std::move(t), nullptr);
+    if (ret < 0) {
+      ss << "Error error writing superblock: " << cpp_strerror(tr);
```
Comment (idryomov) [inline src/osd/OSD.cc:3137]:
```suggestion
      ss << "Error writing superblock: " << cpp_strerror(ret);
```

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -3126,6 +3126,19 @@ will start to track new ops received afterwards.";
     scrub_purged_snaps();
   }
 
+  else if (prefix == "reset_purged_snaps_last") {
+    lock_guard l(osd_lock);
+    superblock.purged_snaps_last = 0;
+    ObjectStore::Transaction t;
+    dout(10) << __func__ << " updating superblock" << dendl;
+    write_superblock(cct, superblock, t);
+    int ret = store->queue_transaction(service.meta_ch, std::move(t), nullptr);
```
Comment (idryomov) [inline src/osd/OSD.cc:3135]: This is shadowing a declaration at the top of the method:
```suggestion
    ret = store->queue_transaction(service.meta_ch, std::move(t), nullptr);
```

**Comment by Matan-B** [comment]
> Sounds good.
> 
> What is in scope of this PR is the issue of OSDs not picking up purged snap IDs from the monitors under normal operation. Do we know what is causing that?

Not yet. This issue was never reported (AFAIK) and this PR is proposed specifically to address the single known instance of it.
Hopefully, if it will appear in the future again, we will be able to better investigate the root cause.

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
> This issue was never reported (AFAIK) and this PR is proposed specifically to address the single known instance of it.

I think I saw at least one more report but I can't find it right now.  It was a small cluster suffering from other issues too, so it ended up just being recreated.

**Comment by athanatos** [comment]
@Matan-B It seems pretty simple to add it to the thrashing test, so I think we should.  A command tested in teuthology is vastly easier to recommend to a user than one that isn't.

**Comment by Matan-B** [comment]
@rzarzynski, tests look good.
I don't think it requires additional qa as this command won't be called in any other (non-thrash) tests.

https://pulpito.ceph.com/matan-2023-10-10_12:05:40-rados:thrash-wip-matanb-reset_purged_snaps_last-test-distro-default-smithi/

**Comment by athanatos** [comment]
jenkins test windows

**Comment by athanatos** [comment]
jenkins test make check arm64

**Review by athanatos — APPROVED** [review]
Seems good to merge to me once the checks pass.

**Review by rzarzynski — APPROVED** [review]

**Pull request merged**